File: bilmarknad-mcp/src/bilmarknad_mcp/tradera.py
from __future__ import annotations

import os
import time
import xml.etree.ElementTree as ET
from typing import Any

import httpx

from bilmarknad_mcp.schema import CarListing
from bilmarknad_mcp.soh import apply_soh

API_BASE = "https://api.tradera.com/v3"
SEARCH_URL = f"{API_BASE}/searchservice.asmx/Search"
GET_ITEM_URL = f"{API_BASE}/publicservice.asmx/GetItem"

# Fordon > Bilar on Tradera (override with TRADERA_CAR_CATEGORY_ID).
DEFAULT_CAR_CATEGORY_ID = 10

CACHE_TTL_SEARCH_SEC = 30 * 60
CACHE_TTL_ITEM_SEC = 30 * 60


class TraderaUnavailableError(RuntimeError):
    """Raised when Tradera cannot be reached or credentials are missing."""


def _local_tag(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def _text(value: Any) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    return text or None


def _to_int(value: Any) -> int | None:
    if value is None or value == "":
        return None
    try:
        return int(float(str(value)))
    except (TypeError, ValueError):
        return None


def _to_float(value: Any) -> float | None:
    if value is None or value == "":
        return None
    try:
        return float(str(value))
    except (TypeError, ValueError):
        return None


def _as_list(value: Any) -> list[Any]:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    return [value]


def etree_to_obj(elem: ET.Element) -> Any:
    children = list(elem)
    if not children:
        text = (elem.text or "").strip()
        return text if text else None
    result: dict[str, Any] = {}
    for child in children:
        key = _local_tag(child.tag)
        value = etree_to_obj(child)
        if key in result:
            existing = result[key]
            if isinstance(existing, list):
                existing.append(value)
            else:
                result[key] = [existing, value]
        else:
            result[key] = value
    return result


def parse_xml_response(text: str) -> dict[str, Any]:
    root = ET.fromstring(text)
    obj = etree_to_obj(root)
    return {_local_tag(root.tag): obj}


def _dig(data: dict[str, Any], *keys: str) -> Any:
    current: Any = data
    for key in keys:
        if not isinstance(current, dict):
            return None
        matched = None
        for candidate, value in current.items():
            if _local_tag(candidate) == key or candidate == key:
                matched = value
                break
        current = matched
    return current


def _parse_attributes(attr_values: Any) -> dict[str, str]:
    attrs: dict[str, str] = {}
    term_values = _dig(attr_values if isinstance(attr_values, dict) else {}, "TermAttributeValues", "TermAttributeValue")
    for entry in _as_list(term_values):
        if not isinstance(entry, dict):
            continue
        name = _text(entry.get("Name"))
        values = entry.get("Values")
        value = None
        if isinstance(values, dict):
            value = _text(values.get("string"))
        elif isinstance(values, list):
            value = _text(values[0]) if values else None
        else:
            value = _text(values)
        if not name or not value:
            continue
        key = name.lower()
        if key in {"condition", "skick"}:
            attrs["condition"] = value
        elif key in {"mobile_brand", "märke", "brand", "make"}:
            attrs["brand"] = value
        elif key in {"mobile_model", "modell", "model"}:
            attrs["model"] = value
    return attrs


def _parse_image_urls(image_links: Any) -> list[str]:
    urls: list[str] = []
    for entry in _as_list(image_links):
        if isinstance(entry, dict):
            url = _text(entry.get("Url") or entry.get("url"))
        else:
            url = _text(entry)
        if url:
            urls.append(url)
    return urls


def _parse_item(item: dict[str, Any], detailed: bool = False) -> dict[str, Any]:
    attrs = _parse_attributes(item.get("AttributeValues"))
    seller = item.get("Seller") if isinstance(item.get("Seller"), dict) else {}
    parsed = {
        "item_id": _to_int(item.get("Id") or item.get("ItemId")) or 0,
        "short_description": _text(item.get("ShortDescription") or item.get("Title")) or "",
        "long_description": _text(item.get("LongDescription")),
        "category_id": _to_int(item.get("CategoryId")) or 0,
        "category_name": _text(item.get("CategoryName")),
        "seller_id": _to_int(item.get("SellerId")) or 0,
        "seller_alias": _text(item.get("SellerAlias")),
        "seller_city": _text(item.get("SellerCity") or seller.get("City")),
        "start_price": _to_int(item.get("StartPrice")),
        "buy_it_now_price": _to_int(item.get("BuyItNowPrice")),
        "current_bid": _to_int(item.get("MaxBid") or item.get("CurrentBid")),
        "bid_count": _to_int(item.get("TotalBids")) or 0,
        "start_date": _text(item.get("StartDate")),
        "end_date": _text(item.get("EndDate")),
        "thumbnail_url": _text(item.get("ThumbnailLink")),
        "image_urls": _parse_image_urls(item.get("ImageLinks")),
        "item_type": _text(item.get("ItemType")) or "Auction",
        "item_url": _text(item.get("ItemUrl")),
        "attributes": attrs,
    }
    if detailed:
        parsed["seller_rating"] = _to_float(item.get("SellerDsrAverage") or seller.get("TotalRating"))
    return parsed


def _price_sek(item: dict[str, Any]) -> int | None:
    for key in ("buy_it_now_price", "current_bid", "start_price"):
        value = item.get(key)
        if value is not None and value > 0:
            return int(value)
    return None


def parse_tradera_item(raw: dict[str, Any], detailed: bool = False) -> CarListing:
    item = _parse_item(raw, detailed=detailed)
    attrs = item.get("attributes") or {}
    image_urls = item.get("image_urls") or []
    thumbnail = item.get("thumbnail_url")
    listing = CarListing(
        source="tradera",
        id=str(item["item_id"]),
        title=item["short_description"],
        make=attrs.get("brand"),
        model=attrs.get("model"),
        price_sek=_price_sek(item),
        location=item.get("seller_city"),
        dealer_name=item.get("seller_alias"),
        url=item.get("item_url") or f"https://www.tradera.com/item/{item['item_id']}",
        image_url=thumbnail or (image_urls[0] if image_urls else None),
        published_at=item.get("start_date"),
        raw=item,
    )
    apply_soh(
        listing,
        item.get("short_description"),
        item.get("long_description"),
        source="tradera_detail" if detailed else "tradera_search",
    )
    return listing


def _extract_search_items(parsed: dict[str, Any]) -> list[dict[str, Any]]:
    items = _dig(parsed, "SearchResult", "Items")
    if items is None:
        items = _dig(parsed, "Items")
    if items is None:
        return []
    if isinstance(items, dict):
        if any(key in items for key in ("Id", "ItemId", "ShortDescription", "Title")):
            return [items]
        nested = items.get("Item")
        if nested is not None:
            return [entry for entry in _as_list(nested) if isinstance(entry, dict)]
    return [entry for entry in _as_list(items) if isinstance(entry, dict)]


def _extract_item(parsed: dict[str, Any]) -> dict[str, Any] | None:
    item = _dig(parsed, "Item")
    if isinstance(item, dict):
        return item
    for value in parsed.values():
        if isinstance(value, dict) and any(key in value for key in ("Id", "ItemId", "ShortDescription", "Title")):
            return value
    return None


class _TimedCache:
    def __init__(self):
        self._entries: dict[str, tuple[float, Any]] = {}

    def get(self, key: str, ttl_sec: int) -> Any | None:
        entry = self._entries.get(key)
        if entry is None:
            return None
        stored_at, value = entry
        if time.monotonic() - stored_at > ttl_sec:
            self._entries.pop(key, None)
            return None
        return value

    def set(self, key: str, value: Any) -> None:
        self._entries[key] = (time.monotonic(), value)


class TraderaClient:
    def __init__(
        self,
        client: httpx.Client | None = None,
        app_id: str | None = None,
        app_key: str | None = None,
        car_category_id: int | None = None,
    ):
        self._client = client
        self._owns = client is None
        self._app_id = app_id or os.environ.get("TRADERA_APP_ID", "5572")
        self._app_key = app_key or os.environ.get("TRADERA_APP_KEY", "81974dd3-404d-456e-b050-b030ba646d6a")
        raw_category = car_category_id
        if raw_category is None:
            env_category = os.environ.get("TRADERA_CAR_CATEGORY_ID")
            raw_category = int(env_category) if env_category else DEFAULT_CAR_CATEGORY_ID
        self._car_category_id = raw_category
        self._cache = _TimedCache()

    def _get_client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(timeout=30.0, headers={"User-Agent": "bilmarknad-mcp/0.2"})
        return self._client

    def _auth_params(self) -> dict[str, str]:
        if not self._app_id or not self._app_key:
            raise TraderaUnavailableError("TRADERA_APP_ID and TRADERA_APP_KEY are required")
        return {"appId": str(self._app_id), "appKey": str(self._app_key)}

    def _get_xml(self, url: str, params: dict[str, Any]) -> dict[str, Any]:
        client = self._get_client()
        response = client.get(url, params={**self._auth_params(), **params})
        if response.status_code >= 400:
            raise TraderaUnavailableError(f"Tradera API returned {response.status_code}")
        return parse_xml_response(response.text)

    def search(
        self,
        q: str | None = None,
        rows: int = 20,
        page: int = 1,
        order_by: str = "Relevance",
        category_id: int | None = None,
    ) -> list[CarListing]:
        query = (q or "").strip() or "bil"
        category = self._car_category_id if category_id is None else category_id
        cache_key = f"search:{query}:{category}:{order_by}:{rows}:{page}"
        cached = self._cache.get(cache_key, CACHE_TTL_SEARCH_SEC)
        if cached is not None:
            return cached

        parsed = self._get_xml(
            SEARCH_URL,
            {
                "query": query,
                "categoryId": str(category),
                "orderBy": order_by,
                "pageNumber": str(page),
            },
        )
        listings = [parse_tradera_item(item) for item in _extract_search_items(parsed)]
        self._cache.set(cache_key, listings[:rows])
        return listings[:rows]

    def get_listing(self, listing_id: str) -> CarListing | None:
        cache_key = f"item:{listing_id}"
        cached = self._cache.get(cache_key, CACHE_TTL_ITEM_SEC)
        if cached is not None:
            return cached

        parsed = self._get_xml(GET_ITEM_URL, {"itemId": str(listing_id)})
        raw_item = _extract_item(parsed)
        if not raw_item:
            return None
        listing = parse_tradera_item(raw_item, detailed=True)
        self._cache.set(cache_key, listing)
        return listing

    def close(self) -> None:
        if self._owns and self._client is not None:
            self._client.close()

File: internal/bilmarknad/tradera/tradera.go
package tradera

import (
	"context"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/httputil"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/schema"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/soh"
)

const (
	apiBase           = "https://api.tradera.com/v3"
	searchURL         = apiBase + "/searchservice.asmx/Search"
	getItemURL        = apiBase + "/publicservice.asmx/GetItem"
	defaultCategoryID = 10
	cacheTTL          = 30 * time.Minute
)

// UnavailableError indicates Tradera API failure or missing credentials.
type UnavailableError struct{ Msg string }

func (e UnavailableError) Error() string {
	if e.Msg != "" {
		return e.Msg
	}
	return "tradera unavailable"
}

type timedCache struct {
	mu      sync.Mutex
	entries map[string]cacheEntry
}

type cacheEntry struct {
	at    time.Time
	value any
}

func newTimedCache() *timedCache {
	return &timedCache{entries: map[string]cacheEntry{}}
}

func (c *timedCache) get(key string, ttl time.Duration) (any, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	entry, ok := c.entries[key]
	if !ok || time.Since(entry.at) > ttl {
		delete(c.entries, key)
		return nil, false
	}
	return entry.value, true
}

func (c *timedCache) set(key string, value any) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.entries[key] = cacheEntry{at: time.Now(), value: value}
}

// Client queries Tradera REST v3 XML APIs.
type Client struct {
	httpClient    *http.Client
	owns          bool
	appID         string
	appKey        string
	carCategoryID int
	cache         *timedCache
}

func NewClient(c *http.Client, appID, appKey string, carCategoryID int) *Client {
	if c == nil {
		c = httputil.NewClient()
	}
	if appID == "" {
		appID = os.Getenv("TRADERA_APP_ID")
		if appID == "" {
			appID = "5572"
		}
	}
	if appKey == "" {
		appKey = os.Getenv("TRADERA_APP_KEY")
		if appKey == "" {
			appKey = "81974dd3-404d-456e-b050-b030ba646d6a"
		}
	}
	if carCategoryID == 0 {
		if env := os.Getenv("TRADERA_CAR_CATEGORY_ID"); env != "" {
			if v, err := strconv.Atoi(env); err == nil {
				carCategoryID = v
			}
		}
	}
	if carCategoryID == 0 {
		carCategoryID = defaultCategoryID
	}
	return &Client{
		httpClient:    c,
		owns:          true,
		appID:         appID,
		appKey:        appKey,
		carCategoryID: carCategoryID,
		cache:         newTimedCache(),
	}
}

func localTag(tag string) string {
	if i := strings.LastIndex(tag, "}"); i >= 0 {
		return tag[i+1:]
	}
	return tag
}

func assignKey(m map[string]any, key string, val any) {
	if existing, ok := m[key]; ok {
		switch e := existing.(type) {
		case []any:
			m[key] = append(e, val)
		default:
			m[key] = []any{e, val}
		}
	} else {
		m[key] = val
	}
}

// ParseXMLResponse parses Tradera XML into a nested map keyed by root tag.
func ParseXMLResponse(text string) (map[string]any, error) {
	decoder := xml.NewDecoder(strings.NewReader(text))
	var root map[string]any
	for {
		tok, err := decoder.Token()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		if start, ok := tok.(xml.StartElement); ok {
			val, err := parseElement(&start, decoder)
			if err != nil {
				return nil, err
			}
			root = map[string]any{localTag(start.Name.Local): val}
			break
		}
	}
	return root, nil
}

func parseElement(start *xml.StartElement, decoder *xml.Decoder) (any, error) {
	var children []any
	var textParts []string
	for {
		tok, err := decoder.Token()
		if err != nil {
			return nil, err
		}
		switch t := tok.(type) {
		case xml.StartElement:
			child, err := parseElement(&t, decoder)
			if err != nil {
				return nil, err
			}
			children = append(children, map[string]any{localTag(t.Name.Local): child})
		case xml.EndElement:
			if t.Name.Local == start.Name.Local {
				if len(children) == 0 {
					if len(textParts) > 0 {
						return strings.Join(textParts, ""), nil
					}
					return nil, nil
				}
				result := map[string]any{}
				for _, child := range children {
					for k, v := range child.(map[string]any) {
						assignKey(result, k, v)
					}
				}
				return result, nil
			}
		case xml.CharData:
			s := strings.TrimSpace(string(t))
			if s != "" {
				textParts = append(textParts, s)
			}
		}
	}
}

func asList(v any) []map[string]any {
	if v == nil {
		return nil
	}
	switch typed := v.(type) {
	case []any:
		out := make([]map[string]any, 0, len(typed))
		for _, item := range typed {
			if m, ok := item.(map[string]any); ok {
				out = append(out, m)
			}
		}
		return out
	case map[string]any:
		if _, ok := typed["Id"]; ok {
			return []map[string]any{typed}
		}
		if item, ok := typed["Item"]; ok {
			return asList(item)
		}
		return []map[string]any{typed}
	default:
		return nil
	}
}

func dig(data map[string]any, keys ...string) any {
	current := any(data)
	for _, key := range keys {
		m, ok := current.(map[string]any)
		if !ok {
			return nil
		}
		var matched any
		for k, v := range m {
			if localTag(k) == key || k == key {
				matched = v
				break
			}
		}
		current = matched
	}
	return current
}

func textVal(v any) string {
	if v == nil {
		return ""
	}
	return strings.TrimSpace(fmt.Sprint(v))
}

func toInt(v any) *int {
	if v == nil {
		return nil
	}
	if p, ok := v.(*int); ok {
		return p
	}
	switch n := v.(type) {
	case int:
		return &n
	case float64:
		i := int(n)
		return &i
	default:
		f, err := strconv.ParseFloat(textVal(v), 64)
		if err != nil {
			return nil
		}
		i := int(f)
		return &i
	}
}

func parseAttributes(attrValues any) map[string]string {
	attrs := map[string]string{}
	m, ok := attrValues.(map[string]any)
	if !ok {
		return attrs
	}
	term := dig(m, "TermAttributeValues", "TermAttributeValue")
	for _, entry := range asList(term) {
		name := textVal(entry["Name"])
		values := entry["Values"]
		var value string
		switch v := values.(type) {
		case map[string]any:
			value = textVal(v["string"])
		case []any:
			if len(v) > 0 {
				value = textVal(v[0])
			}
		default:
			value = textVal(v)
		}
		if name == "" || value == "" {
			continue
		}
		key := strings.ToLower(name)
		switch key {
		case "condition", "skick":
			attrs["condition"] = value
		case "mobile_brand", "märke", "brand", "make":
			attrs["brand"] = value
		case "mobile_model", "modell", "model":
			attrs["model"] = value
		}
	}
	return attrs
}

func parseImageURLs(imageLinks any) []string {
	var urls []string
	for _, entry := range asList(imageLinks) {
		if u := textVal(entry["Url"]); u != "" {
			urls = append(urls, u)
		} else if u := textVal(entry["url"]); u != "" {
			urls = append(urls, u)
		}
	}
	return urls
}

func parseItem(item map[string]any, detailed bool) map[string]any {
	attrs := parseAttributes(item["AttributeValues"])
	seller, _ := item["Seller"].(map[string]any)
	imageURLs := parseImageURLs(item["ImageLinks"])
	itemID := toInt(item["Id"])
	if itemID == nil {
		itemID = toInt(item["ItemId"])
	}
	id := 0
	if itemID != nil {
		id = *itemID
	}
	parsed := map[string]any{
		"item_id":           id,
		"short_description": textVal(item["ShortDescription"]),
		"long_description":  textVal(item["LongDescription"]),
		"category_id":       toInt(item["CategoryId"]),
		"category_name":     textVal(item["CategoryName"]),
		"seller_id":         toInt(item["SellerId"]),
		"seller_alias":      textVal(item["SellerAlias"]),
		"seller_city":       coalesceText(textVal(item["SellerCity"]), textVal(seller["City"])),
		"start_price":       toInt(item["StartPrice"]),
		"buy_it_now_price":  toInt(item["BuyItNowPrice"]),
		"current_bid":       coalesceInt(toInt(item["MaxBid"]), toInt(item["CurrentBid"])),
		"bid_count":         toInt(item["TotalBids"]),
		"start_date":        textVal(item["StartDate"]),
		"end_date":          textVal(item["EndDate"]),
		"thumbnail_url":     textVal(item["ThumbnailLink"]),
		"image_urls":        imageURLs,
		"item_type":         coalesceText(textVal(item["ItemType"]), "Auction"),
		"item_url":          textVal(item["ItemUrl"]),
		"attributes":        attrs,
	}
	if parsed["short_description"] == "" {
		parsed["short_description"] = textVal(item["Title"])
	}
	if detailed {
		parsed["seller_rating"] = toInt(item["SellerDsrAverage"])
		if parsed["seller_rating"] == nil {
			parsed["seller_rating"] = toInt(seller["TotalRating"])
		}
	}
	return parsed
}

func coalesceText(values ...string) string {
	for _, v := range values {
		if strings.TrimSpace(v) != "" {
			return v
		}
	}
	return ""
}

func coalesceInt(values ...*int) *int {
	for _, v := range values {
		if v != nil {
			return v
		}
	}
	return nil
}

func priceSEK(item map[string]any) *int {
	for _, key := range []string{"buy_it_now_price", "current_bid", "start_price"} {
		if v := toInt(item[key]); v != nil && *v > 0 {
			return v
		}
	}
	return nil
}

// ParseTraderaItem converts parsed Tradera item data to CarListing.
func ParseTraderaItem(raw map[string]any, detailed bool) schema.CarListing {
	item := parseItem(raw, detailed)
	attrs, _ := item["attributes"].(map[string]string)
	imageURLs, _ := item["image_urls"].([]string)
	thumbnail := textVal(item["thumbnail_url"])
	var imageURL *string
	if thumbnail != "" {
		imageURL = &thumbnail
	} else if len(imageURLs) > 0 {
		imageURL = &imageURLs[0]
	}
	itemID := fmt.Sprint(item["item_id"])
	itemURL := textVal(item["item_url"])
	if itemURL == "" {
		itemURL = "https://www.tradera.com/item/" + itemID
	}
	listing := schema.CarListing{
		Source:      "tradera",
		ID:          itemID,
		Title:       textVal(item["short_description"]),
		Make:        strPtr(attrs["brand"]),
		Model:       strPtr(attrs["model"]),
		PriceSEK:    priceSEK(item),
		Location:    strPtr(textVal(item["seller_city"])),
		DealerName:  strPtr(textVal(item["seller_alias"])),
		URL:         &itemURL,
		ImageURL:    imageURL,
		PublishedAt: strPtr(textVal(item["start_date"])),
		Raw:         item,
	}
	source := "tradera_search"
	if detailed {
		source = "tradera_detail"
	}
	soh.Apply(&listing, source, textVal(item["short_description"]), textVal(item["long_description"]))
	return listing
}

func strPtr(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

func extractSearchItems(parsed map[string]any) []map[string]any {
	items := dig(parsed, "SearchResult", "Items")
	if items == nil {
		items = dig(parsed, "Items")
	}
	return asList(items)
}

func extractItem(parsed map[string]any) map[string]any {
	if item, ok := dig(parsed, "Item").(map[string]any); ok {
		return item
	}
	for _, v := range parsed {
		if m, ok := v.(map[string]any); ok {
			if _, ok := m["Id"]; ok {
				return m
			}
			if _, ok := m["ItemId"]; ok {
				return m
			}
		}
	}
	return nil
}

func (c *Client) authParams() (map[string]string, error) {
	if c.appID == "" || c.appKey == "" {
		return nil, UnavailableError{Msg: "TRADERA_APP_ID and TRADERA_APP_KEY are required"}
	}
	return map[string]string{"appId": c.appID, "appKey": c.appKey}, nil
}

func (c *Client) getXML(ctx context.Context, url string, params map[string]string) (map[string]any, error) {
	auth, err := c.authParams()
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	q := req.URL.Query()
	for k, v := range auth {
		q.Set(k, v)
	}
	for k, v := range params {
		q.Set(k, v)
	}
	req.URL.RawQuery = q.Encode()
	resp, err := httputil.DoWithRetry(ctx, c.httpClient, req, "tradera", httputil.DefaultRetryPolicy())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 400 {
		return nil, UnavailableError{Msg: fmt.Sprintf("Tradera API returned %d", resp.StatusCode)}
	}
	return ParseXMLResponse(string(body))
}

func (c *Client) Search(ctx context.Context, q *string, rows, page int, orderBy string) ([]schema.CarListing, error) {
	query := "bil"
	if q != nil && strings.TrimSpace(*q) != "" {
		query = strings.TrimSpace(*q)
	}
	if orderBy == "" {
		orderBy = "Relevance"
	}
	cacheKey := fmt.Sprintf("search:%s:%d:%s:%d:%d", query, c.carCategoryID, orderBy, rows, page)
	if cached, ok := c.cache.get(cacheKey, cacheTTL); ok {
		return cached.([]schema.CarListing), nil
	}
	parsed, err := c.getXML(ctx, searchURL, map[string]string{
		"query":      query,
		"categoryId": strconv.Itoa(c.carCategoryID),
		"orderBy":    orderBy,
		"pageNumber": strconv.Itoa(page),
	})
	if err != nil {
		return nil, err
	}
	items := extractSearchItems(parsed)
	out := make([]schema.CarListing, 0, len(items))
	for _, item := range items {
		out = append(out, ParseTraderaItem(item, false))
	}
	if len(out) > rows {
		out = out[:rows]
	}
	c.cache.set(cacheKey, out)
	return out, nil
}

func (c *Client) GetListing(ctx context.Context, listingID string) (*schema.CarListing, error) {
	cacheKey := "item:" + listingID
	if cached, ok := c.cache.get(cacheKey, cacheTTL); ok {
		l := cached.(schema.CarListing)
		return &l, nil
	}
	parsed, err := c.getXML(ctx, getItemURL, map[string]string{"itemId": listingID})
	if err != nil {
		return nil, err
	}
	raw := extractItem(parsed)
	if raw == nil {
		return nil, nil
	}
	listing := ParseTraderaItem(raw, true)
	c.cache.set(cacheKey, listing)
	return &listing, nil
}

func (c *Client) Close() {
	if c.owns && c.httpClient != nil {
		c.httpClient.CloseIdleConnections()
	}
}
